src/sources/util/websocket/source.rs

```diff
@@ -0,0 +1,238 @@
+                }
+            },
+
+            Some(msg) = ws_source.next() => {
+                // Pongs are sent automatically by tungstenite during reading from the stream.
+                match msg {
+                    Ok(Message::Ping(_)) => {
+                        if let Err(error) = ws_sink.send(Message::Pong(PONG.to_vec())).await {
+                            Err(error)
+                        } else {
+                            Ok(())
+                        }
+                    },
+
+                    Ok(Message::Pong(_)) => {
+                        last_pong = Instant::now();
+                        Ok(())
+                    },
+
+                    Ok(Message::Text(msg_txt)) => {
+                        Ok(handle_text_message(&mut out.clone(), WebSocketEvent{
+                            name: DEFAULT_SOURCE_NAME.to_owned(),
+                            payload: msg_txt,
+                            log_namespace: &params.log_namespace,
+                        }, config.uri.clone(),
+                        ).await)
+                    },
+
+                    Ok(Message::Binary(msg_bytes)) => {
+                        let mut buf: BytesMut = msg_bytes.iter().collect();
+                        match params.decoder.clone().decode(&mut buf)
+                            .map(|maybe_msg| async {
+                                maybe_msg.and_then(|(msg, _)| {
+                                    msg.into_iter().nth(0)
+                                }).and_then(|e| {
+                                    if let Event::Log(log_evt) = e {
+                                        Some(WebSocketEvent{
+                                            name: DEFAULT_SOURCE_NAME.to_owned(),
+                                            payload: log_evt.value().to_string(),
+                                            log_namespace: &params.log_namespace,
+                                        })
+                                    } else {
+                                        warn!("Decoded unsupported event: {:?}", e);
+                                        None
+                                    }
+                                }).map(|evt| async {
+                                    handle_text_message(&mut out.clone(), evt, config.uri.clone()).await
+                                }).ok_or(())?.await;
+                                Ok::<(), ()>(())
+                            }).map_err(|err| {error!("Failed to process binary message: {}", err);}) {
+                                Ok(_) => Ok(()),
+                                Err(e) => {
+                                    error!("Failed to send binary message: {:?}", e);
+                                    Ok(())
+                                }
+                            }
+                    },
+
+                    Ok(Message::Close(_)) => {
+                        info!("Received message: connection closed from server");
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏I think since we are emitting the `WsConnectionShutdown` below which logs a `warn` , I think we should remove this one. WDYT?